common/include/ICamera.hpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _I_CAMERA_HPP_
+#define _I_CAMERA_HPP_
+
+#include <nabla.h>
+#include <iostream>
+#include <cstdio>
+#include <fstream>
+#include <chrono>
+
+#include "camera/ICameraControl.hpp"
+
+namespace nbl::hlsl // TODO: DIFFERENT NAMESPACE
+{
+
+template<ProjectionMatrix T = float64_t4x4>
+class ICamera : public ICameraController<typename T>
+{ 
+public:
+	using base_t = typename ICameraController<typename T>;
+
+	struct Traits
+	{
+		using controller_t = base_t;
+		using projection_t = typename controller_t::projection_t;
+		using gimbal_t = typename controller_t::CGimbal;
+		using controller_virtual_event_t = typename controller_t::CVirtualEvent;
+	};
+
+	ICamera(core::smart_refctd_ptr<typename Traits::gimbal_t>&& gimbal, core::smart_refctd_ptr<typename Traits::projection_t> projection, const float32_t3& target = {0,0,0})
+		: base_t(core::smart_refctd_ptr(gimbal)), m_projection(core::smart_refctd_ptr(projection)) { setTarget(target); }
+	~ICamera() = default;
+
+	inline void setPosition(const float32_t3& position)
+	{
+		const auto* gimbal = base_t::m_gimbal.get();
+		gimbal->setPosition(position);
+		recomputeViewMatrix();
+	}
+
+	inline void setTarget(const float32_t3& position)
+	{
+		m_target = position;
+	
+		const auto* gimbal = base_t::m_gimbal.get();
+		auto localTarget = m_target - gimbal->getPosition();
+
+		// TODO: use gimbal to perform a rotation!
+
+		recomputeViewMatrix();
+	}
```

Comment:
may I suggest a trick to avoid recomputing view matrix all the time?

keep a "dirty flag" / incrementing counter in the gimbals that increments after every gimbal manipulation.

then make the camera track the last counter values of all gimbals during `getViewMatrix`, then you know if the view matrix is stale or not and whether to call `recomputeViewMatrix`

Reply:
Also the position and target setting should happen purely based on gimbal manipulation, its a bit of code duplication and inversion of control for anyone to do it through the camera